src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -329,6 +331,35 @@ private Object readResolve() {
             return getDisplayName();
         }
 
+        @Restricted(NoExternalUse.class) // for Jelly
+        public @CheckForNull String getEnclosingLabel() throws Exception {
+            FlowNode executorStepNode = context.get(FlowNode.class);
+            if (executorStepNode != null) {
+                for (StepExecution exec : executorStepNode.getExecution().getCurrentExecutions(true).get(1, TimeUnit.MINUTES)) {
+                    FlowNode runningNode = exec.getContext().get(FlowNode.class);
+                    if (runningNode != null) {
+                        // See if this step is inside our node {} block, and track the associated label.
+                        boolean match = false;
+                        String label = null;
+                        FlowNodeSerialWalker.EnhancedIterator it = new FlowNodeSerialWalker(runningNode).iterator();
```

Comment:
> Why bother?

Normally I'd be all for keeping it simple but we've already gotten bitten by some of these assumptions.  In this case I'd ask  why force the master to do extra work with every UI paint, when it's trivial to avoid it?  

Also, what forces the entire flowgraph to stay in memory, from what I am seeing, shouldn't it just be head nodes + whatever is in the soft-reference cache of flownodes in the xstream FlowNodeStorage, or am I missing a set of full hard references?